internal/hcsoci/resources.go

```diff
@@ -33,3 +40,120 @@ func NormalizeMemorySize(ctx context.Context, cid string, requestedSizeMB uint64
 	}
 	return actualMB
 }
+
+// NormalizeUVMUpdateResourcesRequest accepts an unmarshalled [spec.UpdateTaskRequest.Resources]
+// and the accompanying annotations and returns the Resources struct with the appropriate fields
+// updated based on the VM resource annotations.
+func NormalizeUVMUpdateResourcesRequest(ctx context.Context, data interface{}, annots map[string]string) (interface{}, error) {
+	// calling this directly in uvm.Update() would cause an import cycle, so the best alternative is to
+	// call from containerd-shim-runhcs
+	return normalizeUpdateResourcesRequest(ctx, false, data, annots)
+}
+
+// NormalizeContainerUpdateResourcesRequest is similar to [NormalizeUVMUpdateResourcesRequest],
+// except it uses container resource annotations.
+func NormalizeContainerUpdateResourcesRequest(ctx context.Context, data interface{}, annots map[string]string) (interface{}, error) {
+	return normalizeUpdateResourcesRequest(ctx, true, data, annots)
+}
+
+func normalizeUpdateResourcesRequest(ctx context.Context, isContainer bool, data interface{}, annots map[string]string) (interface{}, error) {
+	var (
+		memorySizeInMB  = annotations.MemorySizeInMB
+		processorCount  = annotations.ProcessorCount
+		processorLimit  = annotations.ProcessorLimit
+		processorWeight = annotations.ProcessorWeight
+	)
+	if isContainer {
+		memorySizeInMB = annotations.ContainerMemorySizeInMB
+		processorCount = annotations.ContainerProcessorCount
+		processorLimit = annotations.ContainerProcessorLimit
+		processorWeight = annotations.ContainerProcessorWeight
+	}
+
+	// could use [ConvertCPULimits], but that would only work for containers, and wouldn't distinguish between
+	// an annotation being set to "0", and the annotation not being present.
+	// And would still need to write out logic for uVM CPU annotation parsing.
+	//
+	// Don't bother checking if multiple CPU values are set for containers, since hcsTask will check in [updateWCOWResources].
+	switch resources := data.(type) {
+	case *specs.WindowsResources:
+		if x := oci.ParseAnnotationsUint64(ctx, annots, memorySizeInMB, 0); x != 0 {
+			if resources.Memory == nil {
+				resources.Memory = &specs.WindowsMemoryResources{}
+			}
+			if y := resources.Memory.Limit; y != nil && *y > 0 {
+				return nil, fmt.Errorf("conflicting memory limits between resource limit (%d) and %q annotation (%d)", *y, memorySizeInMB, x)
+			}
+			xx := x * memory.MiB
```

Comment:
Can we add a comment that `WindowsMemoryResources` limit is in bytes and the annotation is in mb?